### PR TITLE
ARROW-17844: [C++] Remove atomic shared_ptr compatibility functions

### PR DESCRIPTION
**Pull request by pitrou**

### CODE REVIEW

**Review by westonpace — APPROVED** [review]
Good catch, thanks.

File: cpp/src/arrow/array/array_nested.cc [hunk 1]
```diff
@@ -33,7 +33,6 @@
 #include "arrow/type.h"
 #include "arrow/type_fwd.h"
 #include "arrow/type_traits.h"
-#include "arrow/util/atomic_shared_ptr.h"
```
Comment (westonpace) [inline cpp/src/arrow/array/array_nested.cc:36]: Technically I suppose we should add `#include <atomic>` for IWYU but it is probably transitively included by most standard includes.

Comment (pitrou) [reply cpp/src/arrow/array/array_nested.cc:36]: You're right, let's do it.

**Comment by pitrou** [comment]
CI looks green on my fork (apart from a seemingly unrelated R failure), will merge.

**Pull request merged**